pkgs/build-support/shell-builder.nix

```diff
@@ -0,0 +1,45 @@
+{ lib, stdenv }:
+
+# A special kind of derivation that is only meant to be consumed by the
+# nix-shell.
+{
+  mergeInputs ? [], # a list of derivations whose inputs will be merged
```

Comment:
This does not explain what it means for an input to be "merged". (And it's not clear to me from your description.)

Also, linguistically, `mergeInputs` sounds like a Boolean variable, i.e. whether or not to merge inputs, not "list of things to be merged".

Reply:
How about `inputsFrom`? a list of derivations whose inputs will be made available to the environment